### PR TITLE
Combine reservoir models

### PR DESCRIPTION
**Pull request by AnnaKwa**
This PR makes some changes to the existing `Reservoir` and `ReservoirComputingReadout` classes to allow for easier prediction over multiple subdomains in a single call. 

- `Reservoir.increment_state` can operate on a matrix where the each column is the input from a separate subdomain. This saves iterations if we are to use the same reservoir weights across multiple subdomains. Previously, reservoir state updates only worked if the input and reservoir states were 1d arrays.
- `ReservoirComputingReadout` is refactored to save the linear regression coefficients and intercepts as attributes, rather than an sklearn regressor object. This makes combining multiple readouts more straightforward. Instead of being initialized with an sklearn regressor object, the class is initialized with the `ReadoutHyperparameters` and the sklearn regressor is created and fit when `.fit` is called. The class can be initialized with previously-fit values for these, by default they are None and updated when `ReservoirComputingReadout.fit` is called. An exception is raised if `.fit` is called but these attributes are not None.

This also adds a `CombinedReservoirComputingReadout` class that combines multiple trained readouts into a single readout. Let's say there's N subdomains, each with a reservoir state vector of length s. Unlike the method of combining reservoir prediction across subdomains, each subdomain is assumed to use a different readout, so the input to the combined readout `.predict` is a 1D array of length N x s. The coefficient matrices of N individual readouts are combined into a block diagonal matrix. The intercepts are concatenated into a 1D array of length N x s.

- [x] Tests added

Resolves #<github issues> [JIRA-TAG]

Coverage reports (updated automatically):
- test_unit: [60%](https://output.circle-artifacts.com/output/job/8a6fd204-026b-44c7-8cce-01f6d30b9890/artifacts/0/tmp/coverage/htmlcov-test_unit/index.html)

### CODE REVIEW

File: external/fv3fit/fv3fit/reservoir/readout.py [hunk 1]
```diff
@@ -20,36 +27,116 @@ def square_even_terms(v: np.ndarray, axis=1) -> np.ndarray:
...
+            )
+        linear_regressor = Ridge(**self.hyperparameters.linear_regressor_kwargs)
+        if self.square_half_hidden_state is True:
             res_states = square_even_terms(res_states, axis=1)
-        self.linear_regressor.fit(res_states, output_states)
+        linear_regressor.fit(res_states, output_states)
+        self.coefficients = linear_regressor.coef_
+        self.intercepts = linear_regressor.intercept_
 
-    def predict(self, input):
```
Comment (AnnaKwa) [inline external/fv3fit/fv3fit/reservoir/readout.py:41]: Dims change here because sklearn regressor assumed prediction is on 2D array, but we don't use it anymore.

File: external/fv3fit/fv3fit/reservoir/readout.py [hunk 2]
```diff
@@ -20,36 +27,115 @@ def square_even_terms(v: np.ndarray, axis=1) -> np.ndarray:
...
+            )
+        self.square_half_hidden_state = square_state_settings[0]
+
+    def predict(self, input: np.ndarray):
+        if self.square_half_hidden_state:
+            input = square_even_terms(input, axis=0)
+        return self.coefficients * input + self.intercepts
+
+    @classmethod
+    def load(cls, paths: Sequence[str]) -> "CombinedReservoirComputingReadout":
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/reservoir/readout.py:122]: Is it going to be feasible to configure/pass this a list of paths? Or should the class itself be responsible for organizing and loading sub-paths from one path? I notice there isn't a `save` method on this.

File: external/fv3fit/fv3fit/reservoir/config.py [hunk 3]
```diff
@@ -35,6 +35,10 @@ class ReservoirHyperparameters:
         where all elements are sampled from random uniform distribution
         [-1, 1]. Changing this affects relative weighting of reservoir memory
         versus the most recent state.
+    ncols_state: If state is 2D (each column is a separate subdomain which uses
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/reservoir/config.py:38]: Is it necessary for this and `state_size` to be passed as configuration/yaml arguments, or can they be determined automatically at train/run time from tensor shapes? This seems fairly difficult to configure and I don't see how it gets ignored for non-2D states.

**Review by mcgibbon — COMMENTED** [review]
Is there a training function Reservoir or CombinedReservoir is used with?

File: external/fv3fit/fv3fit/reservoir/model.py [hunk 4]
```diff
@@ -69,7 +69,12 @@ def load(cls, path: str) -> "ReservoirComputingModel":
 
         f = io.BytesIO(mapper[cls._READOUT_NAME])
         readout_components = joblib.load(f)
-        readout = ReservoirComputingReadout(**readout_components)
+        readout_hyperparameters = dacite.from_dict(
+            ReadoutHyperparameters, readout_components.pop("hyperparameters")
+        )
+        readout = ReservoirComputingReadout(
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/reservoir/model.py:75]: Can you move the low-level logic on how to load a ReservoirComputingReadout into a `ReservoirComputingReadout.load` class method? Similarly for `Reservoir`. This will avoid e.g. your duplicated change in `HybridReservoirComputingModel` when these classes are modified.

File: external/fv3fit/tests/reservoir/test_readout.py [hunk 5]
```diff
@@ -27,48 +28,113 @@ def test_square_even_terms(arr, axis, expected):
...
+            linear_regressor_kwargs={}, square_half_hidden_state=False
+        ),
+        coefficients=np.ones(shape=(2, 2)),
+        intercepts=np.zeros(2),
+    )
+    with pytest.raises(ValueError):
+        CombinedReservoirComputingReadout(readouts=[readout_1, readout_2])
+
+
+def test_combined_load(tmpdir):
```
Comment (mcgibbon) [inline external/fv3fit/tests/reservoir/test_readout.py:115]: Can this test be simplified by writing a `CombinedReservoirComputingReadout.dump` method, that automatically determines subdirectory names? Do we expect the ability to combine paths from potentially separate training runs to be used?

File: external/fv3fit/fv3fit/reservoir/config.py (hunk 3 above)
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/config.py:38]: Removed, and instead `reset_state` now takes an arg `input_shape` which will provide this information when it is needed to initialize the reservoir state.

File: external/fv3fit/tests/reservoir/test_readout.py (hunk 5 above)
Comment (AnnaKwa) [reply external/fv3fit/tests/reservoir/test_readout.py:115]: > Do we expect the ability to combine paths from potentially separate training runs to be used?
No, I don't see that happening.

Since training will be in parallel, I think it will be easier to have the training function handle saving all the separate readouts into subdirectories rather than waiting for all the readouts to train before combining into that class to dump.

File: external/fv3fit/fv3fit/reservoir/readout.py (hunk 2 above)
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/readout.py:122]: Good point, configuring a list of >10s of paths would be annoying. Changed to list and read from sorted subdirs in `path`. I didn't include a dump method on this class because I figured the saving of each readout to the specified subdir could be handled by the training function.

Comment (mcgibbon) [reply external/fv3fit/fv3fit/reservoir/readout.py]: The one thing to keep in mind with that is the responsibility for knowing how sub-models are stored at a low level now exists in two places - your code that writes the models, and this code that reads them. Ideally you'd be able to remove that low level detail from the writing code and keep it on this class. One possible way to do something like this would be to use the same `get_combined_reservoir_paths(path, n_reservoirs)` helper function in both places.

File: external/fv3fit/fv3fit/reservoir/model.py [hunk 6]
```diff
@@ -49,38 +43,19 @@ def dump(self, path: str) -> None:
...
-
-        mapper[self._READOUT_NAME] = self.readout.dumps()
-        metadata = {
-            "reservoir_hyperparameters": dataclasses.asdict(
-                self.reservoir.hyperparameters
-            )
-        }
-        mapper[self._METADATA_NAME] = yaml.safe_dump(metadata).encode("UTF-8")
+        self.readout.dump(path)
+        self.reservoir.dump(path)
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/reservoir/model.py:47]: Is it safe to dump these both to the same path? This class doesn't know the low-level details of how these subcomponents are stored, it's very possible for them to re-use the same filenames (especially if someone modifies them after you).

File: external/fv3fit/fv3fit/reservoir/model.py [hunk 7]
```diff
@@ -49,38 +43,19 @@ def dump(self, path: str) -> None:
...
-        metadata = yaml.safe_load(mapper[cls._METADATA_NAME])
-
-        reservoir_hyperparameters = dacite.from_dict(
-            ReservoirHyperparameters, metadata["reservoir_hyperparameters"]
-        )
-
-        return cls(reservoir=Reservoir(reservoir_hyperparameters), readout=readout,)
+        readout = ReservoirComputingReadout.load(path)
+        reservoir = Reservoir.load(path)
+        return cls(reservoir=reservoir, readout=readout,)
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/reservoir/model.py:54]: Thanks!

File: external/fv3fit/fv3fit/reservoir/model.py [hunk 8]
```diff
@@ -49,38 +43,19 @@ def dump(self, path: str) -> None:
...
 
 
 class HybridReservoirComputingModel:
-    _READOUT_NAME = "readout.pkl"
+    _READOUT_NAME = "readout.bin"
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/reservoir/model.py:58]: Is this attribute still used?

File: external/fv3fit/fv3fit/reservoir/readout.py [hunk 9]
```diff
@@ -17,39 +25,175 @@ def square_even_terms(v: np.ndarray, axis=1) -> np.ndarray:
...
     def __init__(
-        self, linear_regressor: Ridge, square_half_hidden_state: bool = False,
+        self,
+        hyperparameters: ReadoutHyperparameters,
+        coefficients: Optional[np.ndarray] = None,
+        intercepts: Optional[np.ndarray] = None,
     ):
-        self.linear_regressor = linear_regressor
-        self.square_half_hidden_state = square_half_hidden_state
+        self.hyperparameters = hyperparameters
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/reservoir/readout.py:71]: Task: Class docstring needs to be updated for new/changed attributes.

File: external/fv3fit/fv3fit/reservoir/model.py (hunk 8 above)
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/model.py:58]: Oops, no. Removed.

File: external/fv3fit/fv3fit/reservoir/readout.py (hunk 2 above)
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/readout.py]: That's a good idea. I just realized this part needs to be redone before merging- right now I don't think it's possible to use `ReservoirComputingModel.dump` for multiple models and save the readouts to the directory structure assumed here. Combined with the other suggestion to not dump readout/reservoir to the same path, the new assumed directory structure when reading and writing multiple subdomain models is
```
dir
|- model_0
   |- readout
       |- weights, etc.
   |- reservoir
       |- weights, etc.
|- model_1
    |- ...
```
`CombinedReservoirComputingReadout.load` will take `dir` as the path arg and search for directories ending in numbers, then open their readout subdirs.

File: external/fv3fit/fv3fit/reservoir/model.py (hunk 6 above)
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/model.py:47]: See above comment- changed to dump to readout/reservoir subdirectories.

File: external/fv3fit/fv3fit/reservoir/model.py [hunk 10]
```diff
@@ -49,39 +43,19 @@ def dump(self, path: str) -> None:
...
-        readout = ReservoirComputingReadout(**readout_components)
-        metadata = yaml.safe_load(mapper[cls._METADATA_NAME])
-
-        reservoir_hyperparameters = dacite.from_dict(
-            ReservoirHyperparameters, metadata["reservoir_hyperparameters"]
-        )
-
-        return cls(reservoir=Reservoir(reservoir_hyperparameters), readout=readout,)
+        readout = ReservoirComputingReadout.load(path)
+        reservoir = Reservoir.load(f"{path}/{cls._RESERVOIR_SUBDIR}")
```
Comment (mcgibbon) [inline external/fv3fit/fv3fit/reservoir/model.py:53]: nit (non-blocking): leave this for another PR, but it feels odd that one of these is working in a subdirectory and the other is not.

**Review by mcgibbon — APPROVED** [review]
LGTM. One comment to think about for later.
